### PR TITLE
Refactor the addon of global hub operator

### PR DESCRIPTION
**Pull request by yanmxa**
## Summary

image

https://github.com/yanmxa/hub-of-hubs/tree/br_global_agent/operator/pkg/controllers/agent

## Related issue(s)

Fixes # https://issues.redhat.com/browse/ACM-15097

## Tests
* [x] Unit/function tests have been added and incorporated into `make unit-tests`.
* [x] Integration tests have been added and incorporated into `make integration-test`.
* [ ] E2E tests have been added and incorporated into `make e2e-test-all`.
* [ ] List other manual tests you have done.

### CODE REVIEW

File: operator/pkg/controllers/crd/crd_controller.go [hunk 1]
```diff
@@ -101,27 +97,16 @@ func (r *CrdController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
...
-		agentController, err := agent.NewAddonController(r.Manager.GetConfig(), r.Manager.GetClient(), r.operatorConfig)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		err = r.Manager.Add(agentController)
-		if err != nil {
-			return ctrl.Result{}, err
-		}
-		r.agentController = agentController
+	_, err = agent.AddGlobalHubAddonController(ctx, r.Manager,
```
Comment (clyang82) [inline operator/pkg/controllers/crd/crd_controller.go:106]: will it be used to deploy global hub agent?

**Review by clyang82 — CHANGES_REQUESTED** [review]

File: operator/pkg/controllers/agent/default_addon_reconciler.go [hunk 2]
```diff
@@ -27,14 +26,134 @@ import (
...
+	defaultAgentStarted = false
+)
+
+type DefaultAgentReconciler struct {
 	client.Client
-	Log logr.Logger
 }
 
-func (r *AddonInstaller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func AddDefaultAgentReconciler(ctx context.Context, mgr ctrl.Manager) error {
```
[the remark this answers is not in the document]
Comment (yanmxa) [reply operator/pkg/controllers/agent/default_addon_reconciler.go:41]: It has been added on the `addon_controller.go` `// +kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=managedclusters,verbs=get;list;watch`. Beside, why do we need the `create,update, and delete` permission?

File: operator/pkg/controllers/hubofhubs/controller.go [hunk 3]
```diff
@@ -711,6 +721,24 @@ func (r *GlobalHubReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		}
 	}
 
+	// start the addon controllers only if the multiclusterglobalhub is ready
+	if !agent.ReadyToEnableAddonManager(ctx, r.client) {
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/controller.go:725]: We should start it only if ACM resources are ready. Need add `config.IsACMResourceReady` judgement

Comment (yanmxa) [reply operator/pkg/controllers/hubofhubs/controller.go]: Done.

File: operator/pkg/controllers/hubofhubs/controller.go [hunk 4]
```diff
@@ -711,10 +721,29 @@ func (r *GlobalHubReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		}
 	}
 
-	if config.IsACMResourceReady() && config.GetAddonManager() != nil {
-		if reconcileErr = utils.TriggerManagedHubAddons(ctx, r.client, config.GetAddonManager()); reconcileErr != nil {
+	// start the addon controllers only if the multiclusterglobalhub is ready
+	if !config.IsACMResourceReady() || !agent.ReadyToEnableAddonManager(ctx, r.client) {
```
Comment (clyang82) [inline operator/pkg/controllers/hubofhubs/controller.go:725]:
```suggestion
		if config.IsACMResourceReady(){ 
		if agent.ReadyToEnableAddonManager(ctx, r.client) {
		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
	} else {
		once.Do(func() {
			// start the addon manager
			if err = agent.StartGlobalHubAddonManager(ctx, r.config, r.client, r.operatorConfig); err != nil {
				log.Fatalw("failed to start the lobal hub addon manager")
			}
		})
		// start the addon controllers
		if err = agent.AddDefaultAgentController(ctx, r.mgr); err != nil {
			return ctrl.Result{}, reconcileErr
		}
		if _, err = agent.AddHostedAgentController(r.mgr); err != nil {
			return ctrl.Result{}, reconcileErr
			return ctrl.Result{}, reconcileErr
		}
		}

		if config.GetAddonManager() != nil {
			if reconcileErr = utils.TriggerManagedHubAddons(ctx, r.client, config.GetAddonManager()); reconcileErr != nil {
				return ctrl.Result{}, reconcileErr
			}
		}
	}
	}
	}
```

**Review by clyang82 — APPROVED** [review]
LGTM

**Pull request merged**